test/unittest/tests/test_assertions.cpp

```diff
@@ -13,39 +13,335 @@
+
+    connection->reconfigure(connection, "");
+    REQUIRE(all_diag_asserts_off(session) == 0);
+}
+
+/* Reconfigure the connection with diagnostic_asserts as an empty list. */
+TEST_CASE("Reconfigure: diagnostic_asserts empty list", "[assertions]")
 {
     ConnectionWrapper conn(DB_HOME);
+    auto connection = conn.getWtConnection();
     WT_SESSION_IMPL *session = conn.createSession();
 
-    // Check that the new session has set up our test fields correctly.
-    expect_no_assertion(session);
+    REQUIRE(all_diag_asserts_off(session) == 0);
+    connection->reconfigure(connection, "diagnostic_asserts=[]");
+    REQUIRE(all_diag_asserts_off(session) == 0);
+}
+
+/* Reconfigure the connection with diagnostic_asserts as a list with invalid item. */
+TEST_CASE("Reconfigure: diagnostic_asserts with invalid item", "[assertions]")
+{
+    ConnectionWrapper conn(DB_HOME);
+    auto connection = conn.getWtConnection();
+    WT_SESSION_IMPL *session = conn.createSession();
 
-    SECTION("Basic WT_ASSERT_ALWAYS tests")
-    {
-        WT_ASSERT_ALWAYS(session, 1 == 2, "Values are not equal!");
-        expect_assertion(session, "Assertion '1 == 2' failed: Values are not equal!");
+    REQUIRE(all_diag_asserts_off(session) == 0);
+    REQUIRE(
+      connection->reconfigure(connection, "diagnostic_asserts=[slow_operation, panic, INVALID]"));
+    REQUIRE(all_diag_asserts_off(session) == 0);
+}
 
-        WT_ASSERT_ALWAYS(session, 1 == 1, "Values are not equal!");
-        expect_no_assertion(session);
-    }
+/* Reconfigure the connection with diagnostic_asserts as a list of valid items. */
+TEST_CASE("Reconfigure: diagnostic_asserts with valid items", "[assertions]")
+{
+    ConnectionWrapper conn(DB_HOME);
+    auto connection = conn.getWtConnection();
+    WT_SESSION_IMPL *session = conn.createSession();
+
+    connection->reconfigure(connection, "diagnostic_asserts=[data_validation, invalid_op, panic]");
+
+    REQUIRE((DIAGNOSTIC_ASSERTS_ENABLED(session, WT_DIAG_DATA_VALIDATION)) == true);
+    REQUIRE((DIAGNOSTIC_ASSERTS_ENABLED(session, WT_DIAG_INVALID_OP)) == true);
+    REQUIRE((DIAGNOSTIC_ASSERTS_ENABLED(session, WT_DIAG_PANIC)) == true);
+    REQUIRE((DIAGNOSTIC_ASSERTS_ENABLED(session, WT_DIAG_CONCURRENT_ACCESS)) == false);
+    REQUIRE((DIAGNOSTIC_ASSERTS_ENABLED(session, WT_DIAG_OUT_OF_ORDER)) == false);
+    REQUIRE((DIAGNOSTIC_ASSERTS_ENABLED(session, WT_DIAG_SLOW_OPERATION)) == false);
+    REQUIRE((DIAGNOSTIC_ASSERTS_ENABLED(session, WT_DIAG_VISIBILITY)) == false);
+}
+
+/* Reconfigure with assertion categories changed from enabled->disabled and vice-versa. */
+TEST_CASE("Reconfigure: Transition cases", "[assertions]")
+{
+    ConnectionWrapper conn(
+      DB_HOME, "create, diagnostic_asserts= [concurrent_access, out_of_order]");
+    auto connection = conn.getWtConnection();
+    WT_SESSION_IMPL *session = conn.createSession();
+
+    REQUIRE((DIAGNOSTIC_ASSERTS_ENABLED(session, WT_DIAG_CONCURRENT_ACCESS)) == true);
```

Comment:
Can we add a check that shows `WT_DIAG_DATA_VALIDATION` is false before the reconfig?

Reply:
[6f9c71c](https://github.com/wiredtiger/wiredtiger/pull/8655/commits/6f9c71ce3a5e976ccb8eab196a45f50ffc260ddb)